### ISSUE TITLE
Fast JS RRB-tree

### ISSUE THREAD

**Issue by paldepind** [hdgarrood/purescript-sequences#31]
Hello @hdgarrood 

I've implemented a [highly optimized immutable JavaScript list/sequence](https://github.com/funkia/list) that I think could be useful to the PureScript community. I'd love to get your feedback. In particular, I don't know what name to publish the package under.

The library is an implementation of a data-structure called relaxed radix balanced trees. RRB-trees have essentially the same complexities as finger tree and I've also experimented with creating a [highly optimized finger tree implementation](https://github.com/paldepind/finger-tree). But, I've found that RRB-trees offer significantly lower constant factors. I have a benchmark suite [here](https://funkia.github.io/list/benchmarks/) that compares the performance of List to native arrays and other JS libraries. I also ran List with your benchmark suite and I've included the results below. As you can see the performance is very good :smile: 

Since the JS library simply is called List the obvious name for the PureScript package would be purescript-list. But that is already the name of something else. Thus I'm wondering, what would be a good name to pick? As the graphs below show the performance of List is better than `Data.Sequence`. So one option would be to replace the current `purescript-sequences` with a wrapper around List. What do you think about that? Alternatively, is there some other name that would "make sense" and follow the module naming conventions?

Btw, @gabejohnson has created the PureScript bindings for List. He had the great idea to make the API identical to the `Data.Array` API. Therefore it can serve as a faster drop-in replacement for `Data.Array`.

Benchmark results:
![append](https://user-images.githubusercontent.com/521604/37921003-d45dd4a6-3128-11e8-9f0b-f02683f2e207.png)
![insert-lots](https://user-images.githubusercontent.com/521604/37921149-3f084d86-3129-11e8-950d-68cc9f4abd42.png)
![map](https://user-images.githubusercontent.com/521604/37921024-e2850b1c-3128-11e8-95a3-ca1419d4c9c4.png)
![fold](https://user-images.githubusercontent.com/521604/37921038-ee60a914-3128-11e8-89cf-cac0cee52123.png)
![filter](https://user-images.githubusercontent.com/521604/37921076-0358fd9e-3129-11e8-8564-407e8d40a312.png)
![traverse](https://user-images.githubusercontent.com/521604/37921088-0dd749c4-3129-11e8-94d4-cbe9ffa70981.png)
![concat-map](https://user-images.githubusercontent.com/521604/37921124-24c7e850-3129-11e8-90f5-c25e9dfe682b.png)
![apply](https://user-images.githubusercontent.com/521604/37921137-34b4a9ba-3129-11e8-9f9a-22331c6b7a49.png)

**Comment by hdgarrood** [comment]
Thanks for asking :) `purescript-arrays` is MIT licensed, right? I’d suggest mentioning which parts came from Data.Array in a readme somewhere, and I’d also suggest reading through the license text and making sure you’re doing everything it says in there if you haven’t already. If I remember correctly, with projects using the MIT license (and probably most other open source ones) you’re supposed to include the original license text in derivative works.

**Comment by hdgarrood** [comment]
Right, yeah. If it was up to me I’d probably prefer `rrb-list` over `rrb-tree` since the fact that it’s really a tree behind the scenes is mostly just an implementation detail?
